src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1498 @@
+        if (r < 0)
+                return r;
+
+        r = sd_event_add_signal(m->event, NULL, SIGTERM, NULL, NULL);
+        if (r < 0)
+                return r;
+
+        r = sd_event_add_signal(m->event, NULL, SIGRTMIN+18, sigrtmin18_handler, NULL);
+        if (r < 0)
+                return r;
+
+        r = sd_event_add_memory_pressure(m->event, NULL, NULL, NULL);
+        if (r < 0)
+                log_debug_errno(r, "Failed allocate memory pressure event source, ignoring: %m");
+
+        r = sd_bus_default_system(&m->bus);
+        if (r < 0)
+                return r;
+
+        m->notify_fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (m->notify_fd < 0)
+                return -errno;
+
+        (void) mkdir_parents_label(sa.un.sun_path, 0755);
+        (void) sockaddr_un_unlink(&sa.un);
+
+        if (bind(m->notify_fd, &sa.sa, SOCKADDR_UN_LEN(sa.un)) < 0)
+                return -errno;
+
+        r = setsockopt_int(m->notify_fd, SOL_SOCKET, SO_PASSCRED, true);
+        if (r < 0)
+                return r;
+
+        r = sd_event_add_io(m->event, &m->notify_event_source, m->notify_fd, EPOLLIN,
+                            manager_on_notify, m);
+        if (r < 0)
+                return r;
+
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+static int manager_discover_images(Manager *m, TargetClass class) {
+        _cleanup_hashmap_free_ Hashmap *images = NULL;
+        Image *image;
+        int r;
+
+        images = hashmap_new(&image_hash_ops);
+        if (!images)
+                return log_oom();
+
+        r = image_discover((ImageClass) class, NULL, images);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enumerate images: %m");
+
+        HASHMAP_FOREACH(image, images) {
+                _cleanup_(target_unrefp) Target *t = NULL;
+
+                r = target_new(class, image->name, m, &t);
+                if (r < 0)
+                        return r;
+
+                t->path = strdup(image->path);
+                if (!t->path)
+                        return log_oom();
+
+                t->image_type = image->type;
+        }
+
+        return 0;
+}
+
+static int manager_enumerate_components(Manager *m) {
+        _cleanup_close_pair_ int pipe[2] = PIPE_EBADF;
+        _cleanup_(sigkill_waitp) pid_t pid = 0;
+        _cleanup_fclose_ FILE *f = NULL;
+        Target *t;
+        int r;
+
+        r = safe_fork_full("(sd-sysupdate)",
+                           (int[]) { -EBADF, pipe[1], STDERR_FILENO },
+                           NULL, 0,
+                           FORK_RESET_SIGNALS|FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|
+                           FORK_REARRANGE_STDIO|FORK_WAIT|FORK_LOG|FORK_REOPEN_LOG,
+                           &pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                /* Child */
+                execl(sysupdate_binary_path(), "systemd-sysupdate", "components", NULL);
```

Comment:
we should really add json output to this command, and use that. so that the normal output can be changed as needed withut this being an API break.